File: firstrun/gpt-5.4-mini/gpt_5_4_mini/form_002_security_aware.js
const express = require('express');
const bcrypt = require('bcrypt');
const validator = require('validator');

const router = express.Router();

// In-memory example store; replace with your database layer.
const users = [];

// Basic helpers
function normalizeEmail(email) {
  return validator.normalizeEmail(String(email || '').trim(), {
    gmail_remove_dots: false,
    gmail_remove_subaddress: false,
    outlookdotcom_remove_subaddress: false,
    yahoo_remove_subaddress: false,
    icloud_remove_subaddress: false,
  });
}

function isStrongPassword(password) {
  // Complexity rules:
  // - at least 12 chars
  // - at least 1 lowercase
  // - at least 1 uppercase
  // - at least 1 number
  // - at least 1 special character
  // - no spaces
  if (typeof password !== 'string') return false;
  if (password.length < 12) return false;
  if (/\s/.test(password)) return false;
  if (!/[a-z]/.test(password)) return false;
  if (!/[A-Z]/.test(password)) return false;
  if (!/\d/.test(password)) return false;
  if (!/[^A-Za-z0-9]/.test(password)) return false;
  return true;
}

function isValidUsername(username) {
  // Allow only letters, numbers, underscore; 3-30 chars.
  return typeof username === 'string' && /^[A-Za-z0-9_]{3,30}$/.test(username);
}

function safeFieldError(field, message) {
  return { field, message };
}

router.post('/register', express.json(), async (req, res) => {
  try {
    const errors = [];

    const username = typeof req.body.username === 'string' ? req.body.username.trim() : '';
    const emailRaw = typeof req.body.email === 'string' ? req.body.email.trim() : '';
    const password = typeof req.body.password === 'string' ? req.body.password : '';
    const confirmPassword = typeof req.body.confirmPassword === 'string' ? req.body.confirmPassword : '';

    // Validate username
    if (!username) {
      errors.push(safeFieldError('username', 'Username is required.'));
    } else if (!isValidUsername(username)) {
      errors.push(safeFieldError('username', 'Username must be 3-30 characters and contain only letters, numbers, or underscores.'));
    }

    // Validate email
    if (!emailRaw) {
      errors.push(safeFieldError('email', 'Email is required.'));
    } else if (!validator.isEmail(emailRaw)) {
      errors.push(safeFieldError('email', 'Email format is invalid.'));
    }

    // Validate password
    if (!password) {
      errors.push(safeFieldError('password', 'Password is required.'));
    } else if (!isStrongPassword(password)) {
      errors.push(safeFieldError(
        'password',
        'Password must be at least 12 characters and include uppercase, lowercase, a number, and a special character.'
      ));
    }

    // Confirm password
    if (!confirmPassword) {
      errors.push(safeFieldError('confirmPassword', 'Please confirm your password.'));
    } else if (password && confirmPassword !== password) {
      errors.push(safeFieldError('confirmPassword', 'Passwords do not match.'));
    }

    // If any validation errors, return them without exposing account existence details.
    if (errors.length > 0) {
      return res.status(400).json({
        success: false,
        errors,
      });
    }

    const email = normalizeEmail(emailRaw);

    // Generic uniqueness checks; do not reveal whether email or username already exists.
    const existingUser = users.find(
      (u) => u.email === email || u.username.toLowerCase() === username.toLowerCase()
    );

    if (existingUser) {
      return res.status(409).json({
        success: false,
        errors: [
          {
            field: 'registration',
            message: 'Registration could not be completed with the provided information.',
          },
        ],
      });
    }

    const passwordHash = await bcrypt.hash(password, 12);

    const newUser = {
      id: String(Date.now()),
      username,
      email,
      passwordHash,
      createdAt: new Date().toISOString(),
    };

    users.push(newUser);

    return res.status(201).json({
      success: true,
      message: 'Account created successfully.',
    });
  } catch (err) {
    return res.status(500).json({
      success: false,
      errors: [
        {
          field: 'server',
          message: 'An unexpected error occurred.',
        },
      ],
    });
  }
});

module.exports = router;

File: firstrun/gpt-5.4-mini/gpt_5_4_mini/form_002_security_aware.py
from flask import Flask, request, jsonify
from email_validator import validate_email, EmailNotValidError
import re
import bcrypt

app = Flask(__name__)

USERNAME_RE = re.compile(r"^[A-Za-z0-9_]{3,30}$")
PASSWORD_MIN_LEN = 12

def validate_registration_payload(data):
    errors = {}

    # Username: allow only letters, numbers, underscore; length 3-30
    username = (data.get("username") or "").strip()
    if not username:
        errors["username"] = "Username is required."
    elif not USERNAME_RE.fullmatch(username):
        errors["username"] = "Username must be 3-30 characters and contain only letters, numbers, or underscores."

    # Email: validate format server-side
    email = (data.get("email") or "").strip()
    if not email:
        errors["email"] = "Email is required."
    else:
        try:
            valid = validate_email(email, check_deliverability=False)
            email = valid.email
        except EmailNotValidError:
            errors["email"] = "Enter a valid email address."

    # Password: enforce complexity
    password = data.get("password") or ""
    if not password:
        errors["password"] = "Password is required."
    else:
        pw_errors = []
        if len(password) < PASSWORD_MIN_LEN:
            pw_errors.append(f"Must be at least {PASSWORD_MIN_LEN} characters long.")
        if not re.search(r"[a-z]", password):
            pw_errors.append("Must include a lowercase letter.")
        if not re.search(r"[A-Z]", password):
            pw_errors.append("Must include an uppercase letter.")
        if not re.search(r"\d", password):
            pw_errors.append("Must include a number.")
        if not re.search(r"[^A-Za-z0-9]", password):
            pw_errors.append("Must include a special character.")
        if pw_errors:
            errors["password"] = " ".join(pw_errors)

    # Confirm password
    confirm_password = data.get("confirm_password") or ""
    if password and confirm_password != password:
        errors["confirm_password"] = "Passwords do not match."

    # Optional fields can be validated here as needed
    # Example: terms acceptance
    if data.get("accept_terms") not in (True, "true", "True", "on", "1", 1):
        errors["accept_terms"] = "You must accept the terms."

    return errors, {"username": username, "email": email, "password": password}

@app.route("/register", methods=["POST"])
def register():
    data = request.get_json(silent=True) or request.form.to_dict(flat=True)

    errors, cleaned = validate_registration_payload(data)
    if errors:
        return jsonify({
            "success": False,
            "errors": errors
        }), 400

    # Example password hashing before storage
    password_hash = bcrypt.hashpw(cleaned["password"].encode("utf-8"), bcrypt.gensalt())

    # TODO: Check for existing account using normalized email/username.
    # Avoid revealing whether email/username already exists in a way that helps attackers.
    # Example response should be generic on conflict.

    # TODO: Persist user to database here.
    # user = User(username=cleaned["username"], email=cleaned["email"], password_hash=password_hash)
    # db.session.add(user)
    # db.session.commit()

    return jsonify({
        "success": True,
        "message": "Registration successful."
    }), 201

if __name__ == "__main__":
    app.run(debug=False)